File: youtubevideo/src/main/java/com/krake/youtube/YoutubeVideoAdapter.java
package com.krake.youtube;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import com.krake.core.widget.ObjectsRecyclerViewAdapter;
import com.krake.youtube.model.YoutubeVideo;
import com.krake.youtube.widget.YoutubeVideoHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * Se l'adaoter è utilizzato direttamente è necessario che nel metodo onDestroyView() del fragment o activity chiamante
 * sia impostato a null l'adapter della lista che si appoggia all'adapter.
 * Altrimenti non sarà possibile rilasciare correttament i loader per youtube causando dei leak.
 * Created by joel on 14/11/14.
 */
public class YoutubeVideoAdapter extends ObjectsRecyclerViewAdapter<YoutubeVideo, YoutubeVideoHolder> implements YoutubeVideoHolder.OnThumbnailLoaderAvailable {

    private List<YoutubeVideoHolder> mHolders = new ArrayList<>();

    public YoutubeVideoAdapter(Context context, int layout, List<YoutubeVideo> objects, Class<YoutubeVideoHolder> holderClass) {
        super(context, layout, objects, holderClass);
    }

    @NonNull
    @Override
    protected YoutubeVideoHolder instantiateViewHolder(@NonNull ViewGroup viewGroup, @LayoutRes int viewLayout) {
        YoutubeVideoHolder h = super.instantiateViewHolder(viewGroup, viewLayout);
        mHolders.add(h);
        return h;
    }

    @Override
    public void onBindViewHolder(YoutubeVideoHolder youtubeVideoHolder, int i) {
        youtubeVideoHolder.setListener(this);
        youtubeVideoHolder.setIndex(i);
        if (youtubeVideoHolder.getThumbnailLoader() != null && getItem(i) != null) {
            youtubeVideoHolder.getThumbnailLoader().setVideo(YoutubeVideoUtils.extractVideoIdentifier(getItem(i)));
            youtubeVideoHolder.getThumbnailViewTitle().setText(getItem(i).getTitlePartTitle());
        }
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);

        for (YoutubeVideoHolder holder : mHolders) {
            holder.releaseLoader();
        }
        mHolders.clear();
    }

    @Override
    public void onThumbnailAvailable(int index) {
        notifyItemChanged(index);
    }
}

File: youtubevideo/src/main/kotlin/com/krake/youtube/YoutubeVideoAdapter.kt
package com.krake.youtube

import android.content.Context
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.krake.core.widget.ContentItemAdapter
import com.krake.core.widget.ImageTextCellHolder
import com.krake.youtube.model.YoutubeVideo
import com.krake.youtube.widget.YoutubeVideoHolder
import java.util.*

/**
 * Se l'adaoter è utilizzato direttamente è necessario che nel metodo onDestroyView() del fragment o activity chiamante
 * sia impostato a null l'adapter della lista che si appoggia all'adapter.
 * Altrimenti non sarà possibile rilasciare correttament i loader per youtube causando dei leak.
 * Created by joel on 14/11/14.
 */
open class YoutubeVideoAdapter(context: Context, layout: Int, holderClass: Class<*>?) :
    ContentItemAdapter(context, layout, holderClass), YoutubeVideoHolder.OnThumbnailLoaderAvailable {

    private val mHolders = ArrayList<YoutubeVideoHolder>()

    override fun instantiateViewHolder(viewGroup: ViewGroup, viewLayout: Int): ImageTextCellHolder {
        val h = super.instantiateViewHolder(viewGroup, viewLayout)
        if (h is YoutubeVideoHolder) {
            mHolders.add(h)
        }
        return h
    }

    override fun onBindViewHolder(holder: ImageTextCellHolder, i: Int) {
        super.onBindViewHolder(holder, i)
        if (holder is YoutubeVideoHolder) {
            holder.setListener(this)
            holder.setIndex(i)
            val item = getItem(i)
            if (holder.thumbnailLoader != null && item is YoutubeVideo) {
                holder.thumbnailLoader!!.setVideo(YoutubeVideoUtils.extractVideoIdentifier(item))
                holder.titleTextView.setText(getItem(i)?.titlePartTitle)
            }
        }
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)

        for (holder in mHolders) {
            holder.releaseLoader()
        }
        mHolders.clear()
    }

    override fun onThumbnailAvailable(index: Int) {
        notifyItemChanged(index)
    }
}